[decisions] Transformations (#4621)
ElektraInitiative/libelektra: doc/decisions/0_drafts/transformations.md

# Transformations

## Problem

Certain plugins, e.g. `rename`, perform extensive transformations on keys and keysets within Elektra.
Those transformations include, but are not limited to:

- Changing the names of the keys inbetween `kdbGet` and `kdbSet` operations
- Changing values back and forth for normalization, e.g. `true` -> `1`, `1` -> `true`

> - Adding and removing metadata (e.g. the `spec` plugin)

Author: Yes that's true, but was the only plugin from the top of my head that changed metadata

Apart from the problems with notifications, the way key name changing plugins work also breaks change tracking in plugins like `dbus`.
This is because key names are read-only when they are contained in a `KeySet`.
In order to change the name of a key, such a plugin has to create a new key with the changed name, and delete the key with the old name.
The `dbus` plugin implements change tracking by checking the 'key needs sync' flag instead of comparing the values.
As new keys by design have the 'key needs sync' flag set, the plugins that implement change tracking via the flag will always erroneously detect transformed keys as changed.

### Observed problems with changing values (normalization)

Suppose we have a plugin that changes the value of a key to the hard-coded value `1` during the storage phase of the `set` operation.
If a plugin does change tracking, this will lead to false positives.

```
user:/limits/openfiles = 1
```

If the user changes the value, e.g. using `kdb set user:/limits/openfiles 23`, plugins will observe the new value `23`.
However, the value-changing plugin will reset that value back to `1`.
So in practice, the configuration has not been changed.

> Plugins relying on change tracking plugins (e.g. notification plugins) will however think that it has.

Author: Yes, if the change tracking is only based on the `keyNeedsSync` flag. If we do value-based change tracking `kdb set user:/limits/openfiles 1` would not lead to a detected change.

### Observed problems with adding and removing metadata

This is a bit of a mixture between changing key names and changing values.

## Constraints

1.
2.
3.

## Assumptions

1. We want to enable some kind of transformations

> 2.

Reviewer: I would also assume that the "false positives" for change tracking that you described are a very minor problem. They can also be easily solved, if the change tracking API provides both the old and new value for any changed key. Then a plugin that wants to avoid false positives can do an additional value comparison to see, if the actual value has changed.

> 3.

Reviewer: ```suggestion
3. There is no reason to modify or delete existing `meta:/` keys.
4. Newly generated `meta:/...` keys can
    - either stay and get permanently stored during `kdbSet`
    - or be written as `meta:/generated/...`.
    The `meta:/generated/...` metakeys are never stored and are automatically removed during `kdbSet` before `storage` is called.
```

I really can't think of a use case for metadata transformations that don't fulfil these assumptions.

## Considered Alternatives

### Enable setting and unsetting of the `keyNeedsSync` flag

Require plugins that rename keys to remove the `keyNeedsSync` flag.
This would require making the 'key sync' APIs public.

> This does not solve the problem of changing values.

Author: True. Added this to the description of the alternative aswell.

### Use change tracking to replace the `keyNeedsSync` flag

Eliminate the need of the `keyNeedsSync` flag by utilizing the planned change tracking API.

> This does not solve the problem of changing values and false positives.

Reviewer: I would actually make it an assumption that `keyNeedsSync` is replaced with the new change tracking implementation.

Author: On second thought, replacing this flag entirely is not a good idea. Detecting changes will always be more compute-intensive than just checking a flag. This might be overkill for plugins that do not really care about the changes.

Reviewer 2: I am not so sure if there is a big difference. strcmp on such little values is very fast. Comparing 4/8 byte values is **maybe** nearly as fast as checking a flag (benchmark needed).

But I agree with @atmaxinger: if we already have an optimization, there should be a good reason to throw it.

Reviewer: Okay maybe making it an assumption is too much. But replacing `keyNeedsSync` should still be an option we explore. However, I think it's mostly independent of what we discuss here.

I brought it up, because the `keyNeedsSync` flag is actually not as straightforward as it seems. It doesn't cover all cases and even in `libelektra-kdb` (which has more information than e.g. plugins) we need to do extra work. IMO `keyNeedsSync` may still be useful for `libelektra-kdb`, but shouldn't be public API and plugins/applications should use the new change tracking instead.

### Store original names and values in metadata.

> For key name transformations require that transformation plugins set a metakey (e.g. `meta:/elektra/runtimename`) with the runtime name before they do any transformations in the `kdbSet` phase.

Reviewer: I'd prefer, if we say plugins must use a special function to transform key names, e.g.:

```c
typedef const char * (*ElektraNameTransform)(const Key *);

int elektraApplyNameTransform (KeySet * ks, ElektraNameTransform transform);
```

This `elektraKeyTransformName` could set the `meta:/` key or mark the key in some other way. That makes things easier, if we discover the `meta:/` isn't enough and also avoids potential mistakes (like changing an existing `meta:/` key). The function could also use "hacky tricks" to avoid creating new keys, if the `KeySet * ks` holds the only reference.

We could have a similar function for the value of a key.

Author: Good idea! I added it.

This _must not_ be done if this metakey already exists, i.e. another plugin already tranformed it beforehand.
This allows notification and change tracking functionality to work determine and work with the runtime name of the key.

A similiar thing was already attempted for values, i.e. `meta:/origvalue`.

> 

Reviewer: We could also introduce a new phase between before/after `storage` exclusively for transformations. Then we can just do a "fake" call to that phase to get back the runtime names for change tracking.

Author: Great idea!

# Transformations

## Problem

> Certain plugins, e.g. `rename`, perform extensive transformations on keys and keysets within Elektra.

Reviewer 2: The extensiveness is not the problem.

```suggestion
Certain plugins, e.g. `rename`, perform transformations on keys and keysets within Elektra.
```

Those transformations include, but are not limited to:

> - Changing the names of the keys inbetween `kdbGet` and `kdbSet` operations

Reviewer 2: Wouldn't this "inbetween" also be applicable to the others (value and metadata)?

```suggestion
- Changing the names of the keys back and forth
```

- Changing values back and forth for normalization, e.g. `true` -> `1`, `1` -> `true`

> - Adding and removing metadata

Reviewer 2: Can you give an example here to make more clear why this is a transformation.

Author: Well I think we can leave out metadata completely from this decision.

While these features are useful, they do create feature-interaction problems.
More specifically, problems have been observed in conjunction with the following (overlapping) types of plugins:

- notification plugins
- plugins that do change tracking

> The problem, in general, can be described as: Which phase of the KDB should be used for notifications/change tracking?

Reviewer 2: I don't understand this sentence. It also seems to me to be more specific and not more general?

Author: I think "representation" is the better word for "phase" here.

> We differentiate between three types:

Reviewer 2: Please avoid "types" here, IMHO it doesn't bring clarity:

```suggestion
We differentiate between:
```

> - _stored_ name/value/metadata: How it is actually stored, i.e. the state returned by and passed to the `storage` plugins.

Reviewer 2: ```suggestion
- _persistent_ name/value/metadata: How it is actually stored, i.e. the state returned by and passed to the `storage` plugins.
```

> - _runtime_ name/value/metadata: How it is at runtime, i.e. what is returned by `kdbGet` and passed to `kdbSet`.

Reviewer 2: ```suggestion
- _transient_ name/value/metadata: How it is at runtime, i.e. what is returned by `kdbGet` and passed to `kdbSet`.
```

Apart from the problems with notifications, the way key name changing plugins work also breaks change tracking in plugins like `dbus`.
This is because key names are read-only when they are contained in a `KeySet`.
In order to change the name of a key, such a plugin has to create a new key with the changed name, and delete the key with the old name.
The `dbus` plugin implements change tracking by checking the 'key needs sync' flag instead of comparing the values.
As new keys by design have the 'key needs sync' flag set, the plugins that implement change tracking via the flag will always erroneously detect transformed keys as changed.

### Observed problems with changing values (normalization)

Suppose we have a plugin that changes the value of a key to the hard-coded value `1` during the storage phase of the `set` operation.
If a plugin does change tracking, this will lead to false positives.

```
user:/limits/openfiles = 1
```

If the user changes the value, e.g. using `kdb set user:/limits/openfiles 23`, plugins will observe the new value `23`.

> However, the value-changing plugin will reset that value back to `1`.

Reviewer 2: Little style suggestion (can be ignored, see Elements of Style by Jr. Strunk, William):

```suggestion
The value-changing plugin, however, will reset that value back to `1`.
```

So in practice, the configuration has not been changed.
Plugins relying on change tracking plugins (e.g. notification plugins) will however think that it has.

### Observed problems with adding and removing metadata

> This is a bit of a mixture between changing key names and changing values.

Reviewer 2: Better to leave this away from this decision if we don't really have an example for it.

## Constraints

1.
2.
3.

## Assumptions

> 1. We want to enable some kind of transformations

Reviewer 2: At least renames are a must-have (so it should be constraints). It could be implemented in storage plugins, though.

> 2. False positives for change tracking algorithms are only a minor problem.

Reviewer 2: Please define false positives. If we notify on every `kdbSet` many keys even though actually nothing ever changed, we have a big problem.

Author: Yes, but that is exactly what happens. But I think I found a workable solution for this.

> 3. There is no reason to modify or delete existing `meta:/` keys.

Reviewer 2: :+1:

4. Newly generated `meta:/...` keys can
   - either stay and get permanently stored during `kdbSet`

>    - or be written as `meta:/generated/...`.

Reviewer 2: Isn't this the internal we already have?

Reviewer: Similar semantics, but a separate prefix of generated data makes more sense IMO. Also `meta:/internal` is split by plugin and each `meta:/internal/<plugin>/foo` should only be used by `<plugin>` and ignored by everything else (at least that's how I understand it).

Eliminate the need of the `keyNeedsSync` flag by utilizing the planned change tracking API.
This does not solve the problem of changing values and false positives.

### Store original names and values in metadata.

For key name transformations require that transformation plugins set a metakey (e.g. `meta:/elektra/runtimename`) with the runtime name before they do any transformations in the `kdbSet` phase.
This _must not_ be done if this metakey already exists, i.e. another plugin already tranformed it beforehand.
This allows notification and change tracking functionality to work determine and work with the runtime name of the key.

A similiar thing was already attempted for values, i.e. `meta:/origvalue`.

### Create an API for transformations

Plugins must use a special function to transform key names, e.g.:

```c
typedef const char * (*ElektraNameTransform)(const Key *);

int elektraApplyNameTransform (KeySet * ks, ElektraNameTransform transform);
```

How the `elektraApplyNameTransform` function marks the original name is an internal implementation detail.
May be as `meta:/` or something else entireley.

Something similar could be done for the value of a key as well.

> 

Reviewer 2: What would such plugins then do differently that the problem does not occur?

Author: If they use this function to transform the name, we internally know that it has an old name. 
This knowledge can be utilized by changetracking and notification plugins to detect false positives.

### Introduce a new phase for transformations

We could also introduce a new phase between before/after storage exclusively for transformations.

> Then we can just do a "fake" call to that phase to get back the runtime names for change tracking.

Reviewer 2: ```suggestion
Then we can just do a "fake" call to that phase to get back the transient names for change tracking.
```

In order to change the name of a key, such a plugin has to create a new key with the changed name, and delete the key with the old name.
The `dbus` plugin implements change tracking by checking the 'key needs sync' flag instead of comparing the values.
As new keys by design have the 'key needs sync' flag set, the plugins that implement change tracking via the flag will always erroneously detect transformed keys as changed.

### Observed problems with changing values (normalization)

Suppose we have a plugin that changes the value of a key to the hard-coded value `1` during the storage phase of the `set` operation.
If a plugin does change tracking, this will lead to false positives.

```
user:/limits/openfiles = 1
```

If the user changes the value, e.g. using `kdb set user:/limits/openfiles 23`, plugins will observe the new value `23`.
The value-changing plugin, however, will reset that value back to `1`.
So in practice, the configuration has not been changed.
Plugins relying on change tracking plugins (e.g. notification plugins) will however think that it has.

## Constraints

1. We want to enable some kind of transformations

> 2. Renaming keys should be possible, at least in storage plugins

Reviewer: Adding "at least in storage plugins" makes this constraint meaningless. Before the end of the `get/storage` phase, `libelektra-kdb` doesn't know about any keys. So storage plugins can do whatever transformations they want to. Any such transformation would be considered part of the semantic translation between the storage format and Elektra.

I guess what we want is "user-defined renaming" without the need to build the whole functionality into every storage plugin separately. I.e. either have separate renaming plugins, or have a generic way through which _all_ storage plugins can support renaming.

A similiar thing was already attempted for values, i.e. `meta:/origvalue`.

### Create an API for transformations

Plugins must use a special function to transform key names, e.g.:

```c
typedef const char * (*ElektraNameTransform)(const Key *);

int elektraApplyNameTransform (KeySet * ks, ElektraNameTransform transform);
```

How the `elektraApplyNameTransform` function marks the original name is an internal implementation detail.
May be as `meta:/` or something else entireley.

Something similar could be done for the value of a key as well.

### Introduce a new phase for transformations

We could also introduce a new phase between before/after storage exclusively for transformations.
Then we can just do a "fake" call to that phase to get back the transient names for change tracking.

### Call value transformation plugin(s) when `keySetValue` / `keySetString` is called

After a value has been set by the user, call the transformation plugin.
We could store those callbacks as metakeys, i.e. `meta:/generated/transformation/value/callback/#1`.

In `keySetRaw` we then call all those callbacks.
We also need to provide a simple API to plugins to register such callbacks for a key.

> As the transformations will be applied before `kdbSet`, this will elimate lots of possible false positives in changetracking.

Reviewer: 1. This should be done for name and value, not just value to be consistent.
2. I don't like that it makes `libelektra-core` more complex.
3. The name `meta:/generated/transformation/value/callback/#1` is very long. Using metakeys with such long names uses a lot of memory compare to the naive solution of adding a linked list 
    ```c
    struct callback {
        callback_fn function;
        struct callback * next;
    }
    ```
    to the key struct.
4. It's a neat idea, but doesn't work. There is a very simple case that isn't covered:
    ```c
    kdbGet (kdb, ks, parent);
    ksAppendKey (ks, keyNew ("user:/foo", KEY_VALUE, "new value", KEY_END);
    kdbSet (kdb, ks, parent);
    ```
    The new key doesn't have the callback and therefore isn't transformed before `kdbSet`.

Author: Good points, but

1. Does this actually make sense for names? IMHO names should be handled as close to the storage plugins as possible. But yes, nothing is preventing us from doing it for names also.
2. This is true, but we already do similar things. i.e. `meta:/binary`.
3. Yes, true. How we implement it is not very important to me.
4. It works to eliminate false positives. 
   - A new key is always a change so that it would get detected correctly. 
   - The transformation plugins could be required to handle those cases too, e.g. if no callback is on that key, do the transformation in `kdbSet` as it is now.

Reviewer: Ad 4: What exactly would the callback do?

To me it sounded like the callback would perform the transformation. But that would mean that a newly added key is never transformed, unless the plugin also checks during `kdbSet`. But then we're back to the same old problem.

To use a concrete example, take the `user:/limits/openfiles` example you gave in the "Problem" section.

How would this solution work in this snippet?

```c
kdbGet (kdb, ks, parent);
ksAppendKey (ks, keyNew ("user:/limits/openfiles", KEY_VALUE, "23", KEY_END);
kdbSet (kdb, ks, parent);
```

IMO this is not solved at all. There is no callback, because it is a new key and if the plugin "simply runs"  during `kdbSet`, we haven't solved anything compare to the problem statement.

Author: Yes, there would be 2 different possibilities where transformations happen:
1. For keys returned by `kdbGet`, i.e. with the callback set --> in the callback
2. new keys, i.e. keys that do not have a callback --> in `kdbSet`

In the case of the example, the transformation on `user:/limits/openfiles` would be in `kdbSet`.

As this is a new key, it doesn't matter that the transformation is done later for notification and changetracking plugins. They should detect this key as modified anyway.

But for existing keys, the transformation would take part long before `kdbSet`, so here we can avoid false positives.

Reviewer: > They should detect this key as modified anyway.

But wasn't that the point of your example? The key value wouldn't actually be modified, because the plugin changes the value back to `1`.

Author: In the case of a new key, there wouldn't be a previous value?

Author: If you are hinting at the case, that `kdbGet` already returns a key `user:/limits/openfiles` and we are overriding this key with `ksAppendKey` with a new key: then we also need to check if the overridden key has a callback and the new one does not. 

If this is the case, then copy the callback to the new case, and execute the callback. So this would be a 3rd possibility where the transformation happens.

Reviewer: > copy the callback to the new key, and execute the callback

So `ksAppendKey` would have to do that?

This solution is getting really complicated with a lot of moving parts...

Author: > So ksAppendKey would have to do that?

Yes, both `ksAppendKey ` and `ksAppend`.

> This solution is getting really complicated with a lot of moving parts...

I don't think it's _that_ complicated, but sure. However, this is the first solution that would really eliminate false positives. I'm always open to new solutions for this problem.

Author: I updated the decision to include what has been discussed here.
